File: packages/raster-plugin-filesystem/ios/Sources/FilesystemPlugin/FilesystemPlugin.swift
import RasterPluginSDK
import Foundation

public enum FilesystemPlugin {
    public static func register() {
        RasterPlugin.register(plugin: "Filesystem", method: "getCacheDirectory") { call in
            let url = FileManager.default.urls(for: .cachesDirectory, in: .userDomainMask).first
            guard let url else {
                call.replyErr(code: "UNAVAILABLE", message: "Cache directory unavailable")
                return
            }
            call.replyOk(["uri": url.absoluteString])
        }

        RasterPlugin.register(plugin: "Filesystem", method: "readText") { call in
            guard let uri = call.args?["uri"] as? String, let url = URL(string: uri) else {
                call.replyErr(code: "INVALID_ARGS", message: "uri is required")
                return
            }
            do {
                let text = try String(contentsOf: url, encoding: .utf8)
                call.replyOk(["text": text])
            } catch {
                call.replyErr(code: "READ_FAILED", message: error.localizedDescription)
            }
        }

        RasterPlugin.register(plugin: "Filesystem", method: "writeText") { call in
            guard let uri = call.args?["uri"] as? String,
                  let text = call.args?["text"] as? String,
                  let url = URL(string: uri) else {
                call.replyErr(code: "INVALID_ARGS", message: "uri and text are required")
                return
            }
            do {
                try text.write(to: url, atomically: true, encoding: .utf8)
                call.replyOk(["uri": url.absoluteString])
            } catch {
                call.replyErr(code: "WRITE_FAILED", message: error.localizedDescription)
            }
        }
    }
}

File: packages/raster-plugin-filesystem/android/src/main/java/dev/raster/plugin/filesystem/FilesystemPlugin.kt
package dev.raster.plugin.filesystem

import android.net.Uri
import dev.raster.plugin.RasterPlugin
import org.json.JSONObject
import java.io.File

object FilesystemPlugin {
    @JvmStatic
    fun register() {
        RasterPlugin.register("Filesystem", "getCacheDirectory") { call ->
            val activity = RasterPlugin.currentActivity()
            if (activity == null) {
                call.replyErr("NO_ACTIVITY", "No activity")
                return@register
            }
            val uri = Uri.fromFile(activity.cacheDir).toString()
            call.replyOk(JSONObject().put("uri", uri))
        }

        RasterPlugin.register("Filesystem", "readText") { call ->
            val activity = RasterPlugin.currentActivity()
            if (activity == null) {
                call.replyErr("NO_ACTIVITY", "No activity")
                return@register
            }
            val uri = call.args?.optString("uri")
            if (uri.isNullOrBlank()) {
                call.replyErr("INVALID_ARGS", "uri is required")
                return@register
            }
            val file = resolveScopedFile(activity.cacheDir, uri)
            if (file == null || !file.exists()) {
                call.replyErr("READ_FAILED", "File not found or out of scope")
                return@register
            }
            runCatching {
                call.replyOk(JSONObject().put("text", file.readText()))
            }.onFailure {
                call.replyErr("READ_FAILED", it.message ?: "read failed")
            }
        }

        RasterPlugin.register("Filesystem", "writeText") { call ->
            val activity = RasterPlugin.currentActivity()
            if (activity == null) {
                call.replyErr("NO_ACTIVITY", "No activity")
                return@register
            }
            val uri = call.args?.optString("uri")
            val text = call.args?.optString("text")
            if (uri.isNullOrBlank() || text == null) {
                call.replyErr("INVALID_ARGS", "uri and text are required")
                return@register
            }
            val file = resolveScopedFile(activity.cacheDir, uri)
                ?: File(activity.cacheDir, "raster-${System.currentTimeMillis()}.txt")
            runCatching {
                file.parentFile?.mkdirs()
                file.writeText(text)
                call.replyOk(JSONObject().put("uri", Uri.fromFile(file).toString()))
            }.onFailure {
                call.replyErr("WRITE_FAILED", it.message ?: "write failed")
            }
        }
    }

    private fun resolveScopedFile(cacheDir: File, uri: String): File? {
        val parsed = Uri.parse(uri)
        val path = parsed.path ?: return null
        val file = File(path)
        val cachePath = cacheDir.canonicalPath
        val targetPath = file.canonicalPath
        return if (targetPath.startsWith(cachePath)) file else null
    }
}